.grit/patterns/valibot_v03.md

```diff
@@ -0,0 +1,105 @@
+# Valibot v0.31 Upgrade
+
+In v0.31, the API was changed to use a `v.pipe` method.
+
+```grit
+
+pattern vb_renames() {
+  or {
+    `custom` => `check`,
+    `BaseSchema` => `GenericSchema`,
+    `Input` => `InferInput`,
+    `Output` => `InferOutput`,
+    `SchemaConfig` => `Config`,
+    `special` => `custom`,
+    `toCustom` => `transform`,
+    `toTrimmed` => `trim`,
+    `toTrimmedEnd` => `trimEnd`,
+    `toTrimmedStart` => `trimStart`
+  }
+}
+
+pattern vb_pipe() {
+  or {
+    `v.string([$args])` => `v.pipe(v.string(), $args)`,
```

Comment:
This should work for any schema function, not just `string`. It is also important that it works with any import alias as well as with direct imports (same applies to the other steps). You can find a list of all schema functions in our [API reference](https://valibot.dev/api/).

Reply:
Sorry I don't really understand the distinction in your first argument - the API is not intuitive enough to me to understand what you're going for / why `v.string([v.email()])` becomes `v.pipe(v.string(), v.email()` but `v.map(v.number(), v.string())` becomes `v.pipe(
  v.map(
    v.number(),
    v.pipe(v.string(), v.url(), v.endsWith('@example.com')), // <-- nested pipe
  )`.
  
I was happy to provide a sample, but don't understand your API intimately.

Reply:
I also think you need to calibrate your expectations. Writing a codemod that handles 100% of all edge cases is usually not worth it - people still need to review the code. The main value is in handling the 90% common cases.

Reply:
Thank you! I will try to have a look later today or in the next few days.